flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/ResultPartitionWithConsumableNotifier.java

```diff
@@ -0,0 +1,174 @@
+ * depends on the PIPELINED vs. BLOCKING characteristic of the result partition. With pipelined
+ * results, receivers are deployed as soon as the first buffer is added to the result partition.
+ * With blocking results on the other hand, receivers are deployed after the partition is finished.
+ */
+public class ResultPartitionWithConsumableNotifier implements ResultPartitionWriter {
+
+	private final TaskActions taskActions;
+
+	private final JobID jobId;
+
+	private final ResultPartitionWriter partitionWriter;
+
+	private final ResultPartitionConsumableNotifier partitionConsumableNotifier;
+
+	private final ResultPartitionType partitionType;
+
+	private boolean hasNotifiedPipelinedConsumers;
+
+	public ResultPartitionWithConsumableNotifier(
+			TaskActions taskActions,
+			JobID jobId,
+			ResultPartitionWriter partitionWriter,
+			ResultPartitionConsumableNotifier partitionConsumableNotifier,
+			ResultPartitionType partitionType) {
+		this.taskActions = checkNotNull(taskActions);
+		this.jobId = checkNotNull(jobId);
+		this.partitionWriter = checkNotNull(partitionWriter);
+		this.partitionConsumableNotifier = checkNotNull(partitionConsumableNotifier);
+		this.partitionType = checkNotNull(partitionType);
+	}
+
+	@Override
+	public BufferBuilder getBufferBuilder() throws IOException, InterruptedException {
+		return partitionWriter.getBufferBuilder();
+	}
+
+	@Override
+	public ResultPartitionID getPartitionId() {
+		return partitionWriter.getPartitionId();
+	}
+
+	@Override
+	public int getNumberOfSubpartitions() {
+		return partitionWriter.getNumberOfSubpartitions();
+	}
+
+	@Override
+	public int getNumTargetKeyGroups() {
+		return partitionWriter.getNumTargetKeyGroups();
+	}
+
+	@Override
+	public void setup() throws IOException {
+		partitionWriter.setup();
+	}
+
+	@Override
+	public boolean addBufferConsumer(BufferConsumer bufferConsumer, int subpartitionIndex) throws IOException {
+		boolean success = partitionWriter.addBufferConsumer(bufferConsumer, subpartitionIndex);
+		if (success) {
+			notifyPipelinedConsumers();
+		}
+
+		return success;
+	}
+
+	@Override
+	public void flushAll() {
+		partitionWriter.flushAll();
+	}
+
+	@Override
+	public void flush(int subpartitionIndex) {
+		partitionWriter.flush(subpartitionIndex);
+	}
+
+	@Override
+	public void finish() throws IOException {
+		partitionWriter.finish();
+
+		notifyPipelinedConsumers();
+	}
+
+	@Override
+	public void fail(Throwable throwable) {
+		partitionWriter.fail(throwable);
+	}
+
+	@Override
+	public void close() throws Exception {
+		partitionWriter.close();
+	}
+
+	/**
+	 * Notifies pipelined consumers of this result partition once.
+	 *
+	 * <p>For PIPELINED {@link org.apache.flink.runtime.io.network.partition.ResultPartitionType}s,
+	 * this will trigger the deployment of consuming tasks after the first buffer has been added.
+	 */
+	private void notifyPipelinedConsumers() {
+		if (!hasNotifiedPipelinedConsumers && partitionType.isPipelined()) {
```

Comment:
can't we check `partitionType.isPipelined` when we create an instance of this class? That way we would not have to make this case distinction here.